lib/charms/prometheus_k8s/v0/prometheus_scrape.py

```diff
@@ -545,84 +546,203 @@ def as_dict_with_promql_labels(self):
-    By default, only the top directory is scanned; for nested scanning, pass `recursive=True`.
+    @staticmethod
+    def _from_file(root_path: str, file_path: Path, topology: JujuTopology) -> List[dict]:
+        """Read a rules file from path, injecting juju topology.
 
-    Args:
-        dir_path: directory containing *.rule files (alert rules without groups).
-        topology: a `JujuTopology` instance.
-        recursive: flag indicating whether to scan for rule files recursively.
+        Args:
+            root_path: path to the root rules folder (used only for generating group name)
+            file_path: path to a *.rule file.
+            topology: juju topology information.
 
-    Returns:
-        a list of prometheus alert rule groups.
-    """
-    alerts = defaultdict(list)
+        Returns:
+            A list of dictionaries representing the rules file, if file is valid (the structure is
+            formed by `yaml.safe_load` of the file); an empty list otherwise.
+        """
+        with file_path.open() as rf:
+            # Load a list of rules from file then add labels and filters
+            try:
+                rule_file = yaml.safe_load(rf)
+
+            except Exception as e:
+                logger.error("Failed to read alert rules from %s: %s", file_path.name, e)
+                return []
+
+            # TODO validate input
+            if AlertRules._is_valid_official_format(rule_file):
+                alert_groups = rule_file["groups"]
+            elif AlertRules._is_valid_lma_format(rule_file):
+                # convert to list of alert groups
+                # group name is made up from the file name
+                alert_groups = [{"name": file_path.stem, "rules": [rule_file]}]
+            else:
+                # invalid/unsupported
+                logger.error("Invalid rules file: %s", file_path.name)
+                return []
+
+            # update rules with additional metadata
+            for alert_group in alert_groups:
+                # update group name with topology and sub-path
+                alert_group["name"] = AlertRules._group_name(
+                    root_path,
+                    str(file_path),
+                    topology,
+                    alert_group["name"],
+                )
+
+                # add "juju_" topology labels
+                for alert_rule in alert_group["rules"]:
+                    if "labels" not in alert_rule:
+                        alert_rule["labels"] = {}
+                    alert_rule["labels"].update(topology.as_dict_with_promql_labels())
+
+                    # insert juju topology filters into a prometheus alert rule
+                    alert_rule["expr"] = topology.render(alert_rule["expr"])
 
-    def _group_name(path) -> str:
+            return alert_groups
+
+    @staticmethod
+    def _group_name(
+        root_path: str, file_path: str, topology: JujuTopology, group_name: str
```

Comment:
`topology` is already a member of the `AlertRules` class. If this method was not static it would not need an additional topology argument thereby simplifying its function signature. It would perhaps be more readable to have a `rename_group` method that did not require the `group_name` argument either. Instead the `rename_group` method could read the group name from a alert rule group, change its name and write it back.

Reply:
This function does not rename anything though - it builds up a string, that ends up being used elsewhere.